### PR TITLE
Make Composite{<:Tuple} work

### LINKED ISSUE #24: Overload `FiniteDifferences.to_vec` for Composite
needed for https://github.com/JuliaDiff/ChainRules.jl/pull/175

### PR DESCRIPTION
**Pull request by oxinabox**
partially closes #24 
I did not do the Composite for structs as that has more open questions about how to do this.
(See #24 for discussion)

### CODE REVIEW

File: src/ChainRulesTestUtils.jl [hunk 1]
```diff
@@ -4,20 +4,36 @@ using ChainRulesCore
...
 using FiniteDifferences
+using FiniteDifferences: to_vec
 using LinearAlgebra
 using Test
 
 const _fdm = central_fdm(5, 1)
 
 export test_scalar, frule_test, rrule_test, isapprox, generate_well_conditioned_matrix
 
+include("to_vec.jl")
```
Comment (oxinabox) [inline src/ChainRulesTestUtils.jl:15]: a follow up PR should break the rest of this out into files

File: src/ChainRulesTestUtils.jl [hunk 2]
```diff
@@ -4,20 +4,36 @@ using ChainRulesCore
...
+function Base.isapprox(d_ad::Composite{<:Tuple}, d_fd::Tuple; kwargs...)
+    return isapprox_vec(d_ad, d_fd; kwargs...)
+end
+function Base.isapprox(
+    d_ad::Composite{P, <:Tuple}, d_fd::Composite{P, <:Tuple}; kwargs...
+) where {P <: Tuple}
+    return isapprox_vec(d_ad, d_fd; kwargs...)
+end
+# Must be for same primal
+Base.isapprox(d_ad::Composite{P}, d_fd::Composite{Q}; kwargs...) where {P,Q} = false
```
Comment (oxinabox) [inline src/ChainRulesTestUtils.jl:35]: this doesn't have a test.

File: test/to_vec.jl [hunk 3]
```diff
@@ -0,0 +1,12 @@
+@testset "to_vec" begin
```
Comment (willtebbutt) [inline test/to_vec.jl:1]: Maybe add a test with a nested composite?

File: test/runtests.jl [hunk 4]
```diff
@@ -4,6 +4,31 @@ using Random
...
+                @test x_comp ≈ x_comp
+
+                @test_throws Exception x_comp ≈  collect(x_tup)
+            end
+
+            @testset "different types" begin
+                # both of these are reasonable diffentials for the `Tuple{Int, Int}` primal
+                @test Composite{Tuple{Int, Int}}(1f0, 2f0) ≈ Composite{Tuple{Int, Int}}(1.0, 2.0)
+
+                # But these have different primals so should not be equal
```
Comment (willtebbutt) [inline test/runtests.jl:25]: It's not clear to me how this comment ties up with the code below. If they're not meant to be equal, then why are they approximately equal?

File: src/ChainRulesTestUtils.jl (hunk 2 above)
Comment (willtebbutt) [inline src/ChainRulesTestUtils.jl:35]:
```suggestion
Base.isapprox(d_ad::Composite{P}, d_fd::Composite{Q}; kwargs...) where {P, Q} = false
```

File: test/runtests.jl (hunk 4 above)
Comment (oxinabox) [reply test/runtests.jl:25]: oh dear that test should say `!≈`  something is wrong there

File: test/runtests.jl [hunk 5]
```diff
@@ -4,6 +4,31 @@ using Random
...
+                @test Composite{Tuple{Int, Int}}(1.0, 2.0) ≈ Composite{Tuple{Float64, Float64}}(1.0, 2.0)
```
Comment (oxinabox) [inline test/runtests.jl:26]:
```suggestion
                @test Composite{Tuple{Int, Int}}(1.0, 2.0) !≈ Composite{Tuple{Float64, Float64}}(1.0, 2.0)
```

**Pull request merged**